lxd/storage/drivers/driver_dir.go

```diff
@@ -0,0 +1,864 @@
+		snapPath := snapshot.MountPoint()
+
+		err := os.MkdirAll(snapPath, 0711)
+		if err != nil {
+			return err
+		}
+
+		revertPaths = append(revertPaths, snapPath)
+
+		// Initialise the snapshot's quota with the parent volume's ID.
+		err = d.initQuota(snapPath, volID)
+		if err != nil {
+			return err
+		}
+
+		ourMount, err := snapshot.Mount(op)
+		if err != nil {
+			return err
+		}
+
+		wrapper := migration.ProgressWriter(op, "fs_progress", fullSnapName)
+		path := shared.AddSlash(snapshot.MountPoint())
+
+		err = rsync.Recv(path, conn, wrapper, volTargetArgs.MigrationType.Features)
+		if err != nil {
+			if ourMount {
+				snapshot.Unmount(op)
+			}
+
+			return err
+		}
+
+		if ourMount {
+			snapshot.Unmount(op)
+		}
+	}
+
+	// Finally the actual volume is sent by sender, so create that last.
+	err = os.MkdirAll(volPath, 0711)
+	if err != nil {
+		return err
+	}
+
+	revertPaths = append(revertPaths, volPath)
+
+	// Initialise the volume's quota using the volume ID.
+	err = d.initQuota(volPath, volID)
+	if err != nil {
+		return err
+	}
+
+	// Set the quota if specified in volConfig or pool config.
+	err = d.setQuota(volPath, volID, vol.Config["size"])
+	if err != nil {
+		return err
+	}
+
+	ourMount, err := vol.Mount(op)
+	if err != nil {
+		return err
+	}
+
+	wrapper := migration.ProgressWriter(op, "fs_progress", vol.Name)
+	path := shared.AddSlash(vol.MountPoint())
+	err = rsync.Recv(path, conn, wrapper, volTargetArgs.MigrationType.Features)
+	if err != nil {
+		if ourMount {
+			vol.Unmount(op)
+		}
+
+		return err
+	}
+
+	if ourMount {
+		vol.Unmount(op)
+	}
+
+	revertPaths = nil
+	return nil
+}
+
+// CreateVolumeFromCopy provides same-pool volume copying functionality.
+func (d *dir) CreateVolumeFromCopy(vol Volume, srcVol Volume, copySnapshots bool, op *operations.Operation) error {
+	if vol.ContentType != ContentTypeFS || srcVol.ContentType != ContentTypeFS {
+		return fmt.Errorf("Content type not supported")
+	}
+
+	srcVolPath := srcVol.MountPoint()
+	volPath := vol.MountPoint()
+
+	// Get the volume ID for the new volumes, which is used to set project quota.
+	volID, err := d.getVolID(vol.Type, vol.Name)
+	if err != nil {
+		return err
+	}
+
+	// Create slice of paths created if revert needed later.
+	revertPaths := []string{}
+	defer func() {
+		// Remove any paths created if we are reverting.
+		for _, path := range revertPaths {
+			d.deleteQuota(path, volID)
+			os.RemoveAll(path)
+		}
+	}()
+
+	err = os.MkdirAll(volPath, 0711)
+	if err != nil {
+		return err
+	}
+
+	revertPaths = append(revertPaths, volPath)
+
+	// Initialise the volume's quota using the volume ID.
+	err = d.initQuota(volPath, volID)
+	if err != nil {
+		return err
+	}
+
+	// Set the quota if specified in volConfig or pool config.
+	err = d.setQuota(volPath, volID, vol.Config["size"])
+	if err != nil {
+		return err
+	}
+
+	bwlimit := d.config["rsync.bwlimit"]
+
+	ourMount, err := srcVol.Mount(op)
```

Comment:
Same comment as above, we're copying a volume within a dir pool, mount/unmount are no-ops on dir, so we can do away with all that logic.

Reply:
@stgraber see comment above, have added Volume.MountTask wrapper for this.